connaisseur/sigstore_validator.py

```diff
@@ -0,0 +1,93 @@
+import json
+import logging
+import re
+import subprocess  # nosec
+
+from connaisseur.crypto import decode_and_verify_ecdsa_key
+from connaisseur.exceptions import (
+    CosignError,
+    CosignTimeout,
+    NotFoundException,
+    ValidationError,
+    UnexpectedCosignData,
+)
+
+
+def get_cosign_validated_digests(image: str, pubkey: str):
+    """
+    Gets and processes cosign validation output for a given `image` and `pubkey`
+    and either returns a list of valid digests or raises a suitable exception
+    in case no valid signature is found or cosign fails.
+    """
+    returncode, stdout, stderr = invoke_cosign(image, pubkey)
+    logging.info(
+        "COSIGN output for image: %s; RETURNCODE: %s; STDOUT: %s; STDERR: %s",
+        image,
+        returncode,
+        stdout,
+        stderr,
+    )
+    digests = []
+    if returncode == 0:
+        for sig in stdout.splitlines():
+            try:
+                sig_data = json.loads(sig)
+                try:
+                    digest = sig_data["Critical"]["Image"].get(
+                        "Docker-manifest-digest", ""
+                    )
+                    if re.match(r"(sha256:)?[0-9A-Fa-f]{64}", digest) is None:
+                        raise Exception(
+                            "digest does not match expected digest pattern."
+                        )
+                except Exception as err:
+                    raise UnexpectedCosignData(
+                        f"could not retrieve valid digest from data received by cosign: {err}"
+                    ) from err
+
+                # remove prefix 'sha256' in case it exists
+                digests.append(digest[7:] if digest[:7] == "sha256:" else digest)
+            except json.JSONDecodeError:
+                logging.info("Non-json signature data from Cosign: %s", sig)
+                pass
```

Comment:
Why do you pass here instead of raising an exception? Given that all other error conditions raise, I'd have expected that here as well

Reply:
Well, I assumed from the cosing code that there might be non-json lines, e.g. when there is a certificate: https://github.com/sigstore/cosign/blob/v0.2.0/cmd/cosign/cli/verify.go#L144
However, it might be a valid question whether we should or not (and just log) raise in the other cases, given that one unexpected json does not mean there is no valid json line. What do you think @annekebr ?

Reply:
True, we shouldn't be raising if there is a signature that can be parsed. I'd go for logging and raising only if `digests` is the empty list in the end. If that is the case there is definitely something odd/new with the cosign output because cosign indicated that it was able to verify the image against the public key but we cannot handle the output correctly.

Reply:
Ah, that also ties into @Starkteetje s comment [above](https://github.com/sse-secure-systems/connaisseur/pull/107/files#r611461294) as we basically reduce everything to the case `digests` is the empty list and would throw a specific error in this case only.

Reply:
So, as far as I see it we have these two options: 
1) If cosign exits with status code one, thus indicating successful signature verification against the given public key, and Connaisseur can find at least one line in the output that is a) valid json and b) this valid json is in the format known to Connaisseur such that it can parse the digest we mutate the admission request to point to the digest we found. We fail with UnexpectedCosignData if cosign returns with 0 but Connaisseur cannot find a single parsable digest in the output.
This would the like in https://github.com/sse-secure-systems/connaisseur/pull/136
2) If cosign exits with status code one, thus indicating successful signature verification against the given public key, but Connaisseur finds at least one line that it cannot parse, we fail with UnexpectedCosignData, thus not admitting the image because we don't trust the cosign output any longer.

@xopham favors option 1. Any other opinions @Starkteetje , @peterthomassen ? 
If option 1 is fine, please approve the above mentioned PR. :)

Reply:
Cosign exits with status code zero, right?

I'd be very strongly in favor of option 2 if there's no reason to expect non JSON data. Trying to accept partial, wrongly formatted data as valid is a great recipe for disaster. If cosign should give us JSON but doesn't give us JSON, either it is broken, there's an attack or we're not compatible. In each of these cases, we should fail

Reply:
Well, we are currently using Cosign v0.2.0. It will only output all valid signatures for a given key in a separate line. However, it seems to be possible that a certificate common name is returned that would probably not be valid json, as mentioned above: https://github.com/sigstore/cosign/blob/v0.2.0/cmd/cosign/cli/verify.go#L144 . Furthermore, it is not prohibited to sign multiple times with a valid public key and one might also sign with the same public key in a future version of cosign that could potentially have a different output format, but might theoretically still pass cosign. I do not think this is a strong case and am not sure if it exists.
However, if there is a valid signature line that fullfils our spec, I feel we are doing exactly what was expected, right?

Reply:
Ah, sorry. I didn't see the link before. I guess that does indeed warrant assuming there might be other stuff in the output. It kind of makes my point still, since e.g. now, we have to trust that cosign doesn't allow linebreaks in `vp.Cert.Subject.CommonName`, which is probably a fair assumption in this instance. Still, I don't like it.
But tbh I also don't see how we could work around that, so I guess, I'm fine with this going forward as is

Reply:
Went for option 2 to be on the safe side.

We probably need to touch this anyway as soon as something changes in the output format; so we probably don't gain much by option in the first place. 

Changed #136